theseus/geometry/so3.py

```diff
@@ -31,10 +51,30 @@ def __repr__(self) -> str:
 
     def __str__(self) -> str:
         with torch.no_grad():
-            raise NotImplementedError
+            return f"SO3(matrix={self.data}), name={self.name})"
 
     def _adjoint_impl(self) -> torch.Tensor:
-        raise NotImplementedError
+        return self.data.clone()
+
+    @staticmethod
+    def _SO3_matrix_check(matrix: torch.Tensor):
+        if matrix.ndim != 3 or matrix.shape[1:] != (3, 3):
+            raise ValueError("3D rotations can only be 3x3 matrices.")
+        _check = (
+            torch.matmul(matrix, matrix.transpose(1, 2)) - torch.eye(3, 3)
+        ).abs().max().item() < SO3.SO3_EPS
+        _check &= (torch.linalg.det(matrix) - 1).abs().max().item() < SO3.SO3_EPS
+
+        if not _check:
+            raise ValueError("Not valid 3D rotations.")
+
+    @staticmethod
+    def _unit_quaternion_check(quaternion: torch.Tensor):
+        if quaternion.ndim != 2 or quaternion.shape[1] != 4:
+            raise ValueError("Quaternions can only be 4-D vectors.")
+
+        if (torch.linalg.norm(quaternion, dim=1) - 1).abs().max().item() >= SO3.SO3_EPS:
+            raise ValueError("Not unit quaternions.")
```

Comment:
How about `f"SO3 only accepts unit quaternions as inputs but norm(quaternion) = {norm:.3f}."`?

Reply:
Agreed. Will replace the text.